Mathlib/Dynamics/SymbolicDynamics/Basic.lean

```diff
@@ -0,0 +1,731 @@
+group element `g : G`, the pattern `p` does not occur in `x` at position `g`.
+
+Intuitively, `forbids F` is the shift space defined by declaring the finite set
+(or family) of patterns `F` to be *forbidden*. A configuration belongs to the subshift if and only
+it avoids all the forbidden patterns. -/
+@[to_additive forbids]
+def mulForbids (F : Set (Pattern A G)) : Set (G → A) :=
+  { x | ∀ p ∈ F, ∀ g : G, ¬ p.mulOccursInAt x g }
+
+attribute [inherit_doc SymbolicDynamics.FullShift.mulForbids] SymbolicDynamics.FullShift.forbids
+
+/-- Shifting a configuration commutes with occurrences of a pattern.
+
+Formally: a pattern `p` occurs in the shifted configuration `mulShift h x` at
+position `g` if and only if it occurs in the original configuration `x` at
+position `g + h`. -/
+@[to_additive occurs_shift
+/-- Shifting a configuration commutes with occurrences of a pattern.
+
+Formally: a pattern `p` occurs in the shifted configuration `shift h x` at
+position `g` if and only if it occurs in the original configuration `x` at
+position `g + h`. -/]
+lemma mulOccurs_mulShift (p : Pattern A G) (x : G → A) (g h : G) :
+    p.mulOccursInAt (mulShift h x) g ↔ p.mulOccursInAt x (g * h) := by
+  constructor <;> intro H u hu <;> simpa [mulShift, mul_assoc] using H u hu
+
+/-- Configurations that avoid a family `F` of patterns are stable under the shift.
+
+Formally: if `x` avoids every `p ∈ F` at every position, then for any `h : G`,
+the shifted configuration `mulShift h x` also avoids every `p ∈ F` at every position. -/
+@[to_additive forbids_shift_invariant
+  /-- Configurations that avoid a family `F` of patterns are stable under the shift.
+
+Formally: if `x` avoids every `p ∈ F` at every position, then for any `h : G`,
+the shifted configuration `shift h x` also avoids every `p ∈ F` at every position. -/]
+lemma mulForbids_shift_invariant (F : Set (Pattern A G)) :
+    ∀ h : G, ∀ x ∈ mulForbids (A := A) (G := G) F, mulShift h x ∈ mulForbids F := by
+  intro h x hx p hp g
+  specialize hx p hp (g * h)
+  -- contraposition
+  contrapose! hx
+  simpa [mulOccurs_mulShift] using hx
+
+end Forbids
+
+section OccursAt
+
+variable {A : Type*} [Inhabited A]
+variable {G : Type*}
+-- We assume right-cancellation throughout this section for uniqueness of preimages under (_ + v).
+variable [Monoid G] [IsRightCancelMul G] [DecidableEq G]
+
+/-- Turn a finite pattern into a configuration, by extending it with
+the default symbol outside its support.
+
+Formally: given a pattern `p` with finite support in `G`, we define a configuration
+`patternToOriginConfig p : G → A` by setting
+* `patternToOriginConfig p i = p.data ⟨i, h⟩` if `i ∈ p.support`,
+* `patternToOriginConfig p i = default` otherwise.
+
+This produces a canonical "completion" of the pattern to a configuration,
+filling all unspecified positions with `default`. -/
+@[to_additive patternToOriginConfig]
+def mulPatternToOriginConfig (p : Pattern A G) : G → A :=
```

Comment:
That's a weird name, because `p` is a `Pattern`, not a `MulPattern` (because there is no `MulPattern`). I'd call this `Pattern.mulExtend` and `Pattern.extend` respectively, to make it possible to use dot notation.